test/common/http/header_map_impl_speed_test.cc

```diff
@@ -0,0 +1,194 @@
+
+/**
+ * Measure the speed of retrieving a header value. The numeric Arg passed by the
+ * BENCHMARK(...) macro call below indicates how many dummy headers this test
+ * will add to the HeaderMapImpl during test setup. The relative performance of
+ * this test for different Arg values will help reveal how the speed of the get()
+ * method depends (or doesn't depend) on the number of other headers in the
+ * HeaderMapImpl.
+ */
+static void HeaderMapImplGet(benchmark::State& state) {
+  const LowerCaseString key("example-key");
+  const std::string value("01234567890123456789");
+  HeaderMapImpl headers;
+  addDummyHeaders(headers, state.range(0));
+  headers.setReference(key, value);
+  size_t successes = 0;
+  for (auto _ : state) {
+    successes += (headers.get(key) != nullptr);
+  }
+  benchmark::DoNotOptimize(successes);
+}
+BENCHMARK(HeaderMapImplGet)->Arg(0)->Arg(1)->Arg(10)->Arg(50);
+
+/**
+ * Measure the retrieval speed of a header for which HeaderMapImpl is expected to
+ * provide special optimizations.
+ */
+static void HeaderMapImplGetInline(benchmark::State& state) {
+  const std::string value("01234567890123456789");
+  HeaderMapImpl headers;
+  addDummyHeaders(headers, state.range(0));
+  headers.insertConnection().value().setReference(value);
+  size_t size = 0;
+  for (auto _ : state) {
+    size += headers.Connection()->value().size();
+  }
+  benchmark::DoNotOptimize(size);
+}
+BENCHMARK(HeaderMapImplGetInline)->Arg(0)->Arg(1)->Arg(10)->Arg(50);
+
+/**
+ * Measure the speed of writing to a header for which HeaderMapImpl is expected to
+ * provide special optimizations.
+ */
+static void HeaderMapImplSetInline(benchmark::State& state) {
+  const std::string value("01234567890123456789");
+  HeaderMapImpl headers;
+  addDummyHeaders(headers, state.range(0));
+  for (auto _ : state) {
+    headers.insertConnection().value().setReference(value);
+  }
+  benchmark::DoNotOptimize(headers.size());
+}
+BENCHMARK(HeaderMapImplSetInline)->Arg(0)->Arg(1)->Arg(10)->Arg(50);
+
+/** Measure the speed of the byteSize() estimation method. */
+static void HeaderMapImplGetByteSize(benchmark::State& state) {
+  HeaderMapImpl headers;
+  addDummyHeaders(headers, state.range(0));
+  uint64_t size = 0;
+  for (auto _ : state) {
+    size += headers.byteSize();
+  }
+  benchmark::DoNotOptimize(size);
+}
+BENCHMARK(HeaderMapImplGetByteSize)->Arg(0)->Arg(1)->Arg(10)->Arg(50);
+
+/** Callback function for use in testing HeaderMap::iterate(). */
+static HeaderMap::Iterate CountingCallback(const HeaderEntry&, void* context) {
+  (*static_cast<size_t*>(context))++;
+  return HeaderMap::Iterate::Continue;
+}
+
+/** Measure the speed of iteration with a lightweight callback. */
+static void HeaderMapImplIterate(benchmark::State& state) {
+  HeaderMapImpl headers;
+  size_t num_callbacks = 0;
+  addDummyHeaders(headers, state.range(0));
+  for (auto _ : state) {
+    headers.iterate(CountingCallback, &num_callbacks);
```

Comment:
fwiw you can still declare this with []{...) syntax rather than having the function above.